dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EventDataValuesValidationHook.java

```diff
@@ -97,16 +98,15 @@ private void validateMandatoryDataValues( Event event, ValidationErrorReporter r
 
         TrackerPreheat preheat = reporter.getBundle().getPreheat();
         ProgramStage programStage = reporter.getBundle().getPreheat().getProgramStage( event.getProgramStage() );
-        final List<String> mandatoryDataElements = programStage.getProgramStageDataElements()
+        final List<MetadataIdentifier> mandatoryDataElements = programStage.getProgramStageDataElements()
             .stream()
             .filter( ProgramStageDataElement::isCompulsory )
-            .map( de -> preheat.getIdSchemes().getDataElementIdScheme()
-                .getIdentifier( de.getDataElement() ) )
+            .map( de -> preheat.getIdSchemes().toMetadataIdentifier( de.getDataElement() ) )
             .collect( Collectors.toList() );
-        List<String> missingDataValue = validateMandatoryDataValue( programStage, event,
+        List<MetadataIdentifier> missingDataValue = validateMandatoryDataValue( programStage, event,
             mandatoryDataElements );
         missingDataValue
-            .forEach( de -> reporter.addError( event, E1303, de ) );
+            .forEach( de -> reporter.addError( event, E1303, de.getIdentifierOrAttributeValue() ) );
```

Comment:
We can either pass the identifier as an arg for the error or we can add a parser for MetadataIdentifier in `TrackerReportUtils`.
Personally, I prefer the second option so we do not risk to forget to get the identifier and make the error difficult to read

Reply:
I am going to address this in https://jira.dhis2.org/browse/DHIS2-12519
I a also like the 2nd option better. The metadata identifier has all the context we need to display the id well 😊